Add optional torch.export.ExportGraphSignature to ONNXProgram

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113686
* #112272
* #113495
* #112289
* __->__ #113477

When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

Commit:
Add optional torch.export.ExportGraphSignature to ONNXProgram

When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input name to the original input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

Commit:
Update on "Add optional torch.export.ExportGraphSignature to ONNXProgram"


When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

Commit:
Update on "Add optional torch.export.ExportGraphSignature to ONNXProgram"


When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

Commit:
Update on "Add optional torch.export.ExportGraphSignature to ONNXProgram"


When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

Commit:
Update on "Add optional torch.export.ExportGraphSignature to ONNXProgram"


When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

Commit:
Update on "Add optional torch.export.ExportGraphSignature to ONNXProgram"


When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

thiagocrepaldi commented:
@BowenBao PTAL

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -638,6 +641,75 @@ def model_proto(self) -> onnx.ModelProto:  # type: ignore[name-defined]
+            )
+        """
+
+        return self._model_signature
```
I feel this is only accurate up till the point for fx-graph in ExportedProgram. Further process inside our exporter that modifies the signature will make this outdated. Unless we apply similar io adapting to this too.

thiagocrepaldi replied:
when user model is `torch.export.ExportedProgram`, we don't have any pass that changes the signature.
The only pass applied to the resulting graph module is `fx_module = passes.InsertTypePromotion(diagnostic_context, fx_module).run()`.

BowenBao replied:
okay, this makes sense.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -638,6 +641,75 @@ def model_proto(self) -> onnx.ModelProto:  # type: ignore[name-defined]
+
+        This information is relevant because ONNX specification often differs from PyTorch's,
+        resulting in a ONNX graph with input and output schema different from the actual PyTorch model implementation.
+        By using the model signature, the users can understand the inputs and outputs differences
```
Due to prior comment, I'm not sure this is accurate.

Commit:
Update on "Add optional torch.export.ExportGraphSignature to ONNXProgram"


When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge
